differential/diffAtlas_utils.R

```diff
@@ -0,0 +1,327 @@
+
+  assaydata
+}
+
+# Make a simple contrasts table
+
+make_exp_contrast_table <- function(analytics){
+  
+  conts <- atlas_contrasts( analytics )
+  
+  conts <- data.frame(do.call(
+    rbind, 
+    lapply(conts, function(x){ 
+      unlist(
+        structure(
+          lapply(slotNames(x), function(y){ 
+            slot_val <- slot(x,y)
+            if (y == 'batch_effects' && length(slot_val) > 0){
+              paste(make.names(unlist(lapply(slot_val, function(z){ z@effect_name }))), collapse=' + ')
+            }else{
+              slot_val
+            }
+          }), 
+          names = slotNames(x))
+      )
+    })
+  ), stringsAsFactors = FALSE)
+  
+  # Make the formula we need for each contrast based on any batch effects present
+  
+  conts$formula <- apply(conts, 1, function(x){
+    formula <- '~ 0 + assay_group_id'
+    if ('batch_effects' %in% names(x) && !is.na(x['batch_effects'])){
+      formula <- paste(formula, x['batch_effects'], sep = ' + ')
+    }
+    formula
+  })
+  
+  conts
+}
+
+# Read a normalised expression table, log fold change table or mean intensities table
+
+read_exp_data_table <- function(expAcc, atlasProcessingDirectory, analytics, experiment, type = NULL, dataFilename = NULL){
+  
+  # Get the platform (array design).
+  arrayDesignPart <- ''
+  if (type != 'raw-counts'){
+    arrayDesignPart <- paste0('_', platform( analytics ))
+  }
+  
+  if (is.null(dataFilename)){
+    # Create the data file name from the 
+    dataFilename <- paste( 
+      expAcc, 
+      arrayDesignPart,
+      "-",
+      type, 
+      '.tsv.undecorated', 
+      sep = "" 
+    )
+  
+    dataFilename <- file.path( atlasProcessingDirectory, dataFilename )
+  }
+  
+  if( !file.exists( dataFilename ) ) {
+    stop( paste( "Cannot find:", dataFilename ) )
+  }
+  
+  cat( paste( "Reading", type, "data from", dataFilename, "...\n" ) )
+  
+  # Read in the normalized data.
+  parsedData <- read.delim( 
+    dataFilename, 
+    header = TRUE, 
+    stringsAsFactors = FALSE,
+    row.names = 1
+  )
+  
+  cat( paste("Successfully read", type, "data.\n") )
+  cat( "Matching data to experiment.\n" )
+  
+  if (type %in% c('normalized-expressions', 'raw-counts')){
+    assayNames <- experiment$AssayName
+    bioRepNames <- experiment$BioRepName
+  }else{
+    
+    # For two color (retrieving fold changes and mean intensities), we have to
+    # account for the the Cy3/5 for each assay. 
+    
+    assayToBiorep <- unique(experiment[,c('AssayNameNoCy', 'BioRepNameNoCy')])
+    
+    assayNames <- as.character(assayToBiorep[,'AssayNameNoCy'])
+    bioRepNames <- as.character(assayToBiorep[,'BioRepNameNoCy'])
+  }
+  
+  parsedData <- parsedData[, assayNames]
+  
+  # Merge technical replicates by mean
+  
+  if (any(duplicated(bioRepNames))){
+    parsedData <- do.call(cbind, lapply(split(data.frame(t(parsedData), check.names = FALSE), bioRepNames), colMeans))
+  }
+  
+  parsedData
```

Comment:
For consistency through all functions
```suggestion
  return ( parsedData )
```